service/matching/tasklist/task_completer.go

```diff
@@ -0,0 +1,209 @@
+	errTaskTypeNotSupported            = errors.New("task type not supported")
+	errTaskNotStarted                  = errors.New("task not started")
+	errDomainIsActive                  = &DomainIsActiveInThisClusterError{Message: "domain is active"}
+	historyServiceOperationRetryPolicy = common.CreateTaskCompleterRetryPolicy()
+)
+
+func newTaskCompleter(tlMgr *taskListManagerImpl, retryPolicy backoff.RetryPolicy) TaskCompleter {
+	return &taskCompleterImpl{
+		domainCache:     tlMgr.domainCache,
+		historyService:  tlMgr.historyService,
+		taskListID:      tlMgr.taskListID,
+		clusterMetadata: tlMgr.clusterMetadata,
+		scope:           tlMgr.scope,
+		logger:          tlMgr.logger,
+		throttleRetry: backoff.NewThrottleRetry(
+			backoff.WithRetryPolicy(retryPolicy),
+			backoff.WithRetryableError(isRetryableError),
+		),
+	}
+}
+
+func (tc *taskCompleterImpl) CompleteTaskIfStarted(ctx context.Context, task *InternalTask) error {
+	op := func() (err error) {
+		domainEntry, err := tc.domainCache.GetDomainByID(task.Event.TaskInfo.DomainID)
+		if err != nil {
+			return fmt.Errorf("unable to fetch domain from cache: %w", err)
+		}
+
+		if _, err = domainEntry.IsActiveIn(tc.clusterMetadata.GetCurrentClusterName()); err == nil {
+			return errDomainIsActive
+		}
+
+		req := &types.HistoryDescribeWorkflowExecutionRequest{
+			DomainUUID: task.Event.TaskInfo.DomainID,
+			Request: &types.DescribeWorkflowExecutionRequest{
+				Domain: task.domainName,
+				Execution: &types.WorkflowExecution{
+					WorkflowID: task.Event.WorkflowID,
+					RunID:      task.Event.RunID,
+				},
+			},
+		}
+
+		workflowExecutionResponse, err := tc.historyService.DescribeWorkflowExecution(ctx, req)
+
+		if errors.As(err, new(*types.EntityNotExistsError)) {
+			return nil
+		} else if err != nil {
+			return fmt.Errorf("unable to fetch workflow execution from the history service: %w", err)
+		}
+
+		if workflowExecutionResponse.WorkflowExecutionInfo == nil {
+			return errWorkflowExecutionInfoIsNil
+		}
+
+		isTaskStarted, err := tc.isTaskStarted(task, workflowExecutionResponse)
+		if err != nil {
+			return fmt.Errorf("unable to determine if task has started: %w", err)
+		}
+
+		if isTaskStarted {
+			return nil
+		}
+
+		tc.scope.
+			Tagged(metrics.DomainTag(task.domainName)).
+			Tagged(metrics.TaskListTag(tc.taskListID.GetName())).
+			IncCounter(metrics.StandbyClusterTasksNotStartedCounterPerTaskList)
+
+		return errTaskNotStarted
+	}
+
+	err := tc.throttleRetry.Do(ctx, op)
+
+	if err == nil {
+		task.Finish(nil)
+
+		tc.scope.
+			Tagged(metrics.DomainTag(task.domainName)).
+			Tagged(metrics.TaskListTag(tc.taskListID.GetName())).
+			IncCounter(metrics.StandbyClusterTasksCompletedCounterPerTaskList)
+
+		return nil
+	}
+
+	if !errors.Is(err, errDomainIsActive) && !errors.Is(err, errTaskNotStarted) {
+		tc.logger.Error("Error completing task on domain's standby cluster", tag.Error(err))
```

Comment:
Nit: Maybe emit a metric here? This seems like a weird case that we might want to monitor

Reply:
context timeouts would fall in this branch which would happen during deployments/restarts. Still fine to log in those cases